gtsam/hybrid/tests/testGaussianMixtureFactor.cpp

```diff
@@ -191,8 +197,421 @@ TEST(GaussianMixtureFactor, Error) {
+
+  // The result should be a bell curve like function
+  // with m1 > m0 close to 3.1333.
+  VectorValues given;
+  given.insert(z, Vector1(3.133));
+  HybridGaussianFactorGraph gfg = hbn.toFactorGraph(given);
+  HybridBayesNet::shared_ptr bn = gfg.eliminateSequential();
+
+  HybridBayesNet expected;
+  expected.emplace_back(
+      new DiscreteConditional(m, "0.325603277954/0.674396722046"));
+
+  EXPECT(assert_equal(expected, *bn));
+}
+
+/* ************************************************************************* */
+/**
+ * Test a model P(x0)P(z0|x0)p(x1|m1)p(z1|x1)p(m1).
+ *
+ * p(x1|m1) has different means and same covariance.
+ *
+ * Converting to a factor graph gives us
+ * P(x0)ϕ(x0)P(x1|m1)ϕ(x1)P(m1)
+ *
+ * If we only have a measurement on z0, then
+ * the probability of x1 should be 0.5/0.5.
+ * Getting a measurement on z1 gives use more information.
+ */
+TEST(GaussianMixtureFactor, TwoStateModel) {
+  double mu0 = 1.0, mu1 = 3.0;
+  auto model = noiseModel::Isotropic::Sigma(1, 2.0);
+
+  DiscreteKey m1(M(1), 2);
+  Key z0 = Z(0), z1 = Z(1), x0 = X(0), x1 = X(1);
+
+  auto c0 = make_shared<GaussianConditional>(x1, Vector1(mu0), I_1x1, model),
+       c1 = make_shared<GaussianConditional>(x1, Vector1(mu1), I_1x1, model);
+
+  auto p_x0 = new GaussianConditional(x0, Vector1(0.0), I_1x1,
+                                      noiseModel::Isotropic::Sigma(1, 1.0));
+  auto p_z0x0 = new GaussianConditional(z0, Vector1(0.0), I_1x1, x0, -I_1x1,
+                                        noiseModel::Isotropic::Sigma(1, 1.0));
+  auto p_x1m1 = new GaussianMixture({x1}, {}, {m1}, {c0, c1});
+  auto p_z1x1 = new GaussianConditional(z1, Vector1(0.0), I_1x1, x1, -I_1x1,
+                                        noiseModel::Isotropic::Sigma(1, 3.0));
+  auto p_m1 = new DiscreteConditional(m1, "0.5/0.5");
+
+  HybridBayesNet hbn;
+  hbn.emplace_back(p_x0);
+  hbn.emplace_back(p_z0x0);
+  hbn.emplace_back(p_x1m1);
+  hbn.emplace_back(p_m1);
+
+  VectorValues given;
+  given.insert(z0, Vector1(0.5));
+
+  {
+    // Start with no measurement on x1, only on x0
+    HybridGaussianFactorGraph gfg = hbn.toFactorGraph(given);
+    HybridBayesNet::shared_ptr bn = gfg.eliminateSequential();
+
+    // Since no measurement on x1, we hedge our bets
+    DiscreteConditional expected(m1, "0.5/0.5");
+
+    EXPECT(assert_equal(expected, *(bn->at(2)->asDiscrete())));
+  }
+
+  {
+    // Now we add a measurement z1 on x1
+    hbn.emplace_back(p_z1x1);
+
+    given.insert(z1, Vector1(2.2));
+    HybridGaussianFactorGraph gfg = hbn.toFactorGraph(given);
+    HybridBayesNet::shared_ptr bn = gfg.eliminateSequential();
+
+    // Since we have a measurement on z2, we get a definite result
+    DiscreteConditional expected(m1, "0.4923083/0.5076917");
+
+    EXPECT(assert_equal(expected, *(bn->at(2)->asDiscrete()), 1e-6));
+  }
+}
+
+/* ************************************************************************* */
+/**
+ * Test a model P(x0)P(z0|x0)p(x1|m1)p(z1|x1)p(m1).
```

Comment:
When fixing, consider using a namespace as a “test fixture” to avoid copy/paste

Reply:
Coolio